src/query/functions-v2/src/scalars/comparison.rs

```diff
@@ -39,89 +39,259 @@ use regex::bytes::Regex;
+    BooleanDomain {
+        has_true: true,
+        has_false: false,
+    }
+}
+
+#[inline(always)]
+fn all_false_domain() -> BooleanDomain {
+    BooleanDomain {
+        has_true: false,
+        has_false: true,
+    }
+}
+
+fn register_string_cmp(registry: &mut FunctionRegistry) {
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "eq",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
-        |lhs, rhs, _| T::upcast_gat(lhs) == T::upcast_gat(rhs),
+        |lhs, rhs, _| lhs == rhs,
     );
-    registry.register_2_arg::<T, T, BooleanType, _, _>(
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "noteq",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
-        |lhs, rhs, _| T::upcast_gat(lhs) != T::upcast_gat(rhs),
+        |lhs, rhs, _| lhs != rhs,
     );
-    registry.register_2_arg::<T, T, BooleanType, _, _>(
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "gt",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
-        |lhs, rhs, _| T::upcast_gat(lhs) > T::upcast_gat(rhs),
+        |lhs, rhs, _| lhs > rhs,
     );
-    registry.register_2_arg::<T, T, BooleanType, _, _>(
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "gte",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
-        |lhs, rhs, _| T::upcast_gat(lhs) >= T::upcast_gat(rhs),
+        |lhs, rhs, _| lhs >= rhs,
     );
-    registry.register_2_arg::<T, T, BooleanType, _, _>(
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "lt",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
-        |lhs, rhs, _| T::upcast_gat(lhs) < T::upcast_gat(rhs),
+        |lhs, rhs, _| lhs < rhs,
     );
-    registry.register_2_arg::<T, T, BooleanType, _, _>(
+    registry.register_2_arg::<StringType, StringType, BooleanType, _, _>(
         "lte",
         FunctionProperty::default(),
         |_, _| FunctionDomain::NoThrow,
-        |lhs, rhs, _| T::upcast_gat(lhs) <= T::upcast_gat(rhs),
+        |lhs, rhs, _| lhs <= rhs,
     );
 }
 
+macro_rules! register_simple_domain_type_cmp {
+    ($registry:ident, $T:ty) => {
+        $registry.register_2_arg::<$T, $T, BooleanType, _, _>(
+            "eq",
+            FunctionProperty::default(),
+            |d1, d2| {
+                if d1.min > d2.max || d1.max < d2.min {
+                    FunctionDomain::Domain(all_false_domain())
+                } else if d1.min == d1.max && d2.min == d2.max {
+                    if d1.min == d2.min {
+                        FunctionDomain::Domain(all_true_domain())
+                    } else {
+                        FunctionDomain::Domain(all_false_domain())
+                    }
+                } else {
+                    FunctionDomain::NoThrow
+                }
```

Comment:
therefore, the following is sufficient:
```suggestion
                if d1.min > d2.max || d1.max < d2.min {
                    FunctionDomain::Domain(all_false_domain())
                } else {
                    FunctionDomain::NoThrow
                }
```